### ISSUE TITLE
Documentation for wrapping xla ops

### ISSUE THREAD

**Issue by gehring** [google/jax#1145]
I ran into a situation which would best be solved by XLA's `ExtractImagePatches` op. Given this doesn't seem to be part of lax yet, I was hoping to implement my own wrapper based off of existing wrappers. Unfortunately, being inexperienced with using XLA directly, the existing wrappers are quite difficult to parse.

It would be helpful to have some documentation explaining all the requirements for implementing a good JAX wrapper and maybe provide a running example going from minimal wrapper to a polished wrapper with all the bells and whistles. Even better would be to have some automatic way of wrapping xla ops (if that is realistically possible). Keep in mind that it is possible that I've missed some obvious things/tools. If this is the case, I sincerely apologize for wasting your time!

**Comment by hawkinsp** [comment]
Just a point of clarification: XLA doesn't have an `ExtractImagePatches` op, but TensorFlow does. When TensorFlow is lowering to XLA, it uses the implementation here:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tf2xla/kernels/extract_image_patches_op.cc

We could easily implement that same operation in JAX as a Python function that calls primitives from `jax.lax`, which are almost 1:1 with the underlying XLA primitives. The implementation would be very similar to the one that TensorFlow uses when lowering to XLA, only in Python instead of C++:

In other words: for this kind of op that can be built as a composition of `lax` primitives, this is actually pretty easy: you just write a Python function in terms of the existing primitives. We should document this better, though!

**Comment by gehring** [comment]
@hawkinsp Thanks for the clarification! I'm a little embarrassed for not looking more carefully. I was searching through the tensorflow repo and found a test for the `ExtractImagePatches` with comments implying it was testing the XLA op. I now see that it was referring to a tensorflow op rather than an XLA primitive.

In the mean time, I went for a more limited implementation to extract patches (i.e., slice + pad + stack), but I might try to implement tensorflow's version and compare the performance. If I do implement the TF version using `jax.lax`, do you think there would be any interest in merging it into JAX?

As for the documentation, if all XLA ops are already wrapped and part of `lax` then I'm not sure documentation for wrapping ops is necessary. Is this the case? Also, it would be nice to have some more documentation about the inner workings of JAX but I don't know enough to tell if it would have any practical value. Though, that is probably a bit out of the scope of this issue.

**Comment by hawkinsp** [comment]
We'd welcome an implementation of `ExtractImagePatches` in JAX! The main question is: where should we put it? I'd be tempted to put it somewhere in `jax.ops`, perhaps even a new package `jax.ops.image`.

I think we certainly need some documentation here, since although this is straightforward ("just write Python code!") you need to know that somehow!

**Comment by gehring** [comment]
I made a prototype of `ExtractImagePatches` and pinged you. Otherwise, I'm not sure if you want to keep this issue open for tracking or not but feel free to close it if you want. I consider my personal issue solved.

**Comment by hawkinsp** [comment]
I think this is now well-covered by `lax.conv_general_dilated_patches`. Closing!
